### ISSUE TITLE
[Lock] Know if the lock was acquired by another process before I acquire it

### ISSUE THREAD

**Issue by Gregoire-M** [symfony/symfony#24430]
| Q                | A
| ---------------- | -----
| Bug report?      | no
| Feature request? | yes
| BC Break report? | no
| RFC?             | no
| Symfony version  | 3.4

Hello,

I use the lock component to avoid race conditions on a doctrine entity updates. I use the `RetryTillSaveStore` store, so if 2 processes wants to update a given a entity, process B will wait that process A has finished its job.

My problem is that if process B has waited for the lock, it means that the entity it has in memory isn't up to date anymore. So I have to call the `refresh` method of the doctrine object manager on my entity. But I don't want to do this refresh call if I've acquired the lock without waiting.

So my request is: When I acquire a lock, I'd like to know if I've waited to obtain it, or if I've acquired it immediately. Is it possible?

TL;DR, here's the pseudo code of my problem:

```php
$lock = $this->factory->createLock($entity->getId());
$lock->acquire(true); // I may wait here several seconds if another process owns the lock
$this->objectManager->refresh($entity); // TODO: guess whether the refresh is required or not?
// do some stuff on $entity and flush it
$lock->release();
```

I don't really know how this behavior could be implemented, the debate is open.

**Comment by nicolas-grekas** [comment]
I think that's really hard to achieve without opening the door to race conditions...

**Comment by Gregoire-M** [comment]
I thought I found a solution by simply calling `$lock->isAcquired()` right after I've created it, before I acquire it:

```php
$lock = $this->factory->createLock($entity->getId());
$isOwnByOtherProcess = $lock->isAcquired();
```

But it return `false` even if a lock is present in the store. (Is this normal by the way?)

Could it be possible to implement a `$lock->isAcquiredByOtherProcess()` or something approaching?

**Comment by stof** [comment]
@Gregoire-M the issue is that the other process may not yet have acquired the lock at the time you call `isAcquired`, but then have done it when you try to acquire it yourselves later. Welcome to the world of concurrent programming.

**Comment by stof** [comment]
Here is an idea for you: first acquire the lock in non-blocking mode. If this throws a LockConflictedException, acquire it again in blocking mode (which will wait and retry thanks to the RetryTillSaveStore) and consider that you need to refresh the entity.

**Comment by jakzal** [comment]
@Gregoire-M have stof's idea solved your problem?

**Comment by Gregoire-M** [comment]
@jakzal It works, but the resulting code looks too complicated to me. For now, I've resigned myself to execute the `refresh` each time.

**Comment by nicolas-grekas** [comment]
Closing as discussed.
